pkg/ctl/install/install.go

```diff
@@ -0,0 +1,435 @@
+	return gitCmd.Run()
+}
+
+func getFluxManifests(params install.TemplateParameters, cs *kubernetes.Clientset) (map[string][]byte, error) {
+	params.AdditionalFluxArgs = []string{"--sync-garbage-collection", "--manifest-generation"}
+	manifests, err := install.FillInTemplates(params)
+	if err != nil {
+		return nil, err
+	}
+	// Verify if the namespace to install Flux already exists and otherwise add a manifest for it
+	_, err = cs.CoreV1().Namespaces().Get(params.Namespace, metav1.GetOptions{})
+	if err == nil {
+		return manifests, nil
+	}
+	if !k8serrors.IsNotFound(err) {
+		return nil, fmt.Errorf("cannot check whether namespace %s exists: %s", params.Namespace, err)
+	}
+	nsTemplate := `---
+apiVersion: v1
+kind: Namespace
+metadata:
+  labels:
+    name: %s
+  name: %s
+`
+	ns := fmt.Sprintf(nsTemplate, params.Namespace, params.Namespace)
+	manifests[namespaceFileName] = []byte(ns)
+	return manifests, nil
+}
+
+func writeFluxManifests(baseDir string, manifests map[string][]byte) error{
+	if err := os.MkdirAll(baseDir, 0700); err != nil {
+		return fmt.Errorf("cannot create Flux manifests directory (%s): %s", baseDir, err)
+	}
+	for fileName, contents := range manifests {
+		fullPath := filepath.Join(baseDir, fileName)
+		if err := ioutil.WriteFile(fullPath, contents, 0600); err != nil {
+			return fmt.Errorf("failed to write Flux manifest file %s: %s", fullPath, err)
+		}
+	}
+	return nil
+}
+
+func runShell(workDir string) error {
+	shell, err := loginshell.Shell()
+	if err != nil {
+		return fmt.Errorf("failed to obtain login shell %s: %s", shell, err)
+	}
+	shellCmd := exec.Cmd{
+		Path:   shell,
+		Dir:    workDir,
+		Stdin:  os.Stdin,
+		Stdout: os.Stdout,
+		Stderr: os.Stderr,
+	}
+	return shellCmd.Run()
+}
+
+func waitForFluxToStart(ctx context.Context, opts *installFluxOpts, restConfig *rest.Config,
+	cs *kubernetes.Clientset) (ssh.PublicKey, error) {
+	fluxSelector := metav1.LabelSelector{
+		MatchExpressions: []metav1.LabelSelectorRequirement{
+			{
+				Key:      "name",
+				Operator: metav1.LabelSelectorOpIn,
+				Values:   []string{"flux"},
+			},
+		},
+	}
+	portforwarder := portforward.PortForward{
+		Labels: fluxSelector,
+		Config: restConfig,
+		Clientset: cs,
+		DestinationPort: 3030,
+		Namespace: opts.templateParams.Namespace,
+	}
+	podDeadline := time.Now().Add(30 * time.Second)
+	for ; time.Now().Before(podDeadline); time.Sleep(1 * time.Second) {
+		err := portforwarder.Start()
+		if err == nil {
+			defer portforwarder.Stop()
+			break
+		}
+		if !strings.Contains(err.Error(), "Could not find running pod for selector") {
+			logger.Warning("Flux is not ready yet (%s), retrying ...", err)
+		}
+	}
+	if time.Now().After(podDeadline) {
+		return ssh.PublicKey{},fmt.Errorf("timed out waiting for flux's pod to be created")
+	}
+	fluxURL := fmt.Sprintf("http://127.0.0.1:%d/api/flux", portforwarder.ListenPort)
+	fluxClient := client.New(http.DefaultClient, transport.NewAPIRouter(), fluxURL, client.Token(""))
```

Comment:
If we need this client just to get flux to generate the SSH keys, I'd rather provide flux with a pre-generated key and stop having to rely on remote state and readiness of flux.

Reply:
To be clear: I'm okay with this as a first cut, but long term I think we should avoid port-forwarding.

Reply:
It serves to guarantee that flux is functioning too. Port forwarding is how fluxctl connects to flux and it works perfectly fine please indicate why you oppose to it. Also, I don't see why you want to generate your own ssh key when flux does it for you

Reply:
It's just a timely operation that we have to wait for, it's very complicated, we shouldn't have introduce it. Happy to continue discuss offline.